README.md
json-proxy
==========

Run HTML5 apps locally and proxy your API calls to remote servers effortlessly
without CORS or JSONP :sunglasses:

Use json-proxy on the command line, as a grunt plugin for `grunt serve` livereloads,
or as middleware inside Express or Connect NodeJS apps.

Forwarding rules match URLs to proxy to remote servers.

Optionally injects custom HTTP request headers when proxying, which is great for
API tokens or authentication credentials during early prototyping.

### Screenshot of the CLI utility

![screenshot](https://raw.github.com/wiki/steve-jansen/json-proxy/screenshot.png)

### Status

[![Build Status](https://travis-ci.org/steve-jansen/json-proxy.svg?branch=master)](https://travis-ci.org/steve-jansen/json-proxy)
[![Code Climate](https://codeclimate.com/github/steve-jansen/json-proxy.png)](https://codeclimate.com/github/steve-jansen/json-proxy)
[![Test Coverage](https://codeclimate.com/github/steve-jansen/json-proxy/coverage.png)](https://codeclimate.com/github/steve-jansen/json-proxy)
[![npm dependencies](https://www.bithound.io/github/steve-jansen/json-proxy/badges/dependencies.svg)](https://www.bithound.io/github/steve-jansen/json-proxy/master/dependencies/npm)

[![NPM](https://nodei.co/npm/json-proxy.png?downloads=true&stars=true)](https://nodei.co/npm/json-proxy/)

### Why write yet another Node.JS proxy??

My shop has much love for HTML5 single page apps that call server-side JSON APIs.
We're pretty open minded about server stacks, so the API might run on a Ruby, .Net,
or Play! app server.  This utility enables our front end UI devs to focus on writing
front-end HTML/CSS/JS goodness and not need to worry about how to provision/build/run
the app server on their local dev machine.

### Installation

For CLI usage:

		npm install -g json-proxy

For Express/Connect middleware:

		npm install json-proxy --save

For Grunt middleware:

		npm install json-proxy --save-dev

### Example Apps

An example of the json-proxy CLI can be run via the following shell scripts:

#### Windows
```
examples/cli/run
```

#### OS X and *nix
```
./examples/cli/run
```

An example of using json-proxy as express middleware can be run via:

```
node examples/middleware/index.js
```

An example of using json-proxy as grunt middleware can be run via:

#### Windows
```
cd examples/grunt
npm install -g grunt
npm install -g bower
npm install ../../../json-proxy
npm install
bower install
grunt serve
```

#### OS X and *nix
```
cd examples/grunt
sudo npm install -g grunt
sudo npm install -g bower
npm install ../../../json-proxy
npm install
bower install
grunt serve
```

### CLI Usage

```bash
  json-proxy [-c configFile] [-p port] [-f proxy forwarding rule]
             [-h header rule] [-html5mode [defaultFile]] [directory]

Examples:
   json-proxy -p 8080 -f "/api=http://server" -f "/=http://localhost:9000" .
   json-proxy -h "X-Forwarded-User=johndoe" /tmp/folder
   json-proxy -c "/tmp/config.json"

By default, looks for a config file at ./json-proxy.json

Environmental variables:
  JSON_PROXY_PORT         see --port
  JSON_PROXY_WEBROOT      directory
  JSON_PROXY_GATEWAY      --gateway
  JSON_PROXY_GATEWAY_AUTH "username:password" credentials for --gateway)

Options:
  -p, --port     The TCP port for the proxy server
  -f, --forward  a forwarding rule (ex. /foo=server/foo)
  -h, --header   a custom request header (ex. iv-user=johndoel)
  -c, --config   a config file
  -g, --gateway  URL for a LAN HTTP proxy to use for forwarding requests
  --html5mode    support AngularJS HTML5 mode by catching 404s
  -?, --help     about this utility
  --version      version info
 ```

### Grunt Usage

For Grunt build files using grunt-contrib-connect v0.8.0 or higher:

```js
livereload: {
  options: {
    middleware: function(connect, options, middlewares) {
      // inject json-proxy to the front of the default middlewares array
      middlewares.unshift(
        require('json-proxy').initialize({
          proxy: {
            forward: {
              '/api/': 'http://api.example.com:8080'
            },
            headers: {
              'X-Forwarded-User': 'John Doe'
            }
          }
        })
      );

      return middlewares;
    }
  }
}
```

You may also maintain the config options in an external file:

```js
livereload: {
  options: {
    middleware: function(connect, options, middlewares) {
      // inject json-proxy to the front of the default middlewares array
      middlewares.unshift(
        require('json-proxy').initialize({file: './myconfig.json' })
      );

      return middlewares;
    }
  }
}
```

For Grunt build files using `lrSnippet` in the livereload task,
place json-proxy before `lrSnippet` in the array of connect middlewares:

```js
livereload: {
    options: {
        middleware: function(connect) {
            return [
		          	require('json-proxy').initialize({
							    proxy: {
							      forward: {
							        '/api': 'http://api.example.com:8080'
							      },
							      headers: {
							      	'X-Forwarded-User': 'John Doe'
							    	}
							    }
							  }), // <-- here
                lrSnippet,
                mountFolder(connect, '.tmp'),
                mountFolder(connect, yeomanConfig.app)
            ];
        }
    }
}
```

### Forwarding Rules

The forwarding rules for proxying support a number of different scenarios.

The forwarding rules use regular expressions in the spirit of nginx rewrite rules.

json-proxy will always preserve the request body (e.g., requests with
`POST`, `PUT`, or `PATCH` verbs).  json-proxy will generally preserve
request headers, except in two situations.  Custom headers in the config will
always clobber the existing value of the same header in the original request.
The proxy will also clobber headers typically used by proxy servers
(e.g., `Via`, `X-Forwarded-For`).

**TIP:** Reserved characters in Regex such as `?` and
[Regex character classes](https://developer.mozilla.org/en/docs/Web/JavaScript/Guide/Regular_Expressions#Using_Special_Characters) like `\d` (digits) and `\s` (whitespace) require escaping with a backslash
in forwaring rules (e.g. `'\?'`, `'\\d'`, `'\\s'`).


#### Forwarding the requested URL "as is" to another server

This scenario is the original use case for json-proxy and works
out of the box without any special syntax. Specify the local URL path
to match and the remote server:

``` js
var config = {
  "forward": {
    "/api/": "http://api.example.com"
  }
};
```

This config would forward requests for `/api/*` to
`http://api.example.com:80/api/*`.


#### Pattern matching a URL


``` js
var config = {
  "forward": {
    "/user/\\d+/email": "http://api.example.com"
  }
};

```

This config would forward requests for `/user/12345/email` to
`http://api.example.biz:80/user/12345/email`.


#### URL Rewriting to delete a base path from the requested URL

You will need to use a regex capture group to delete fragments from the
requested URL.  Use `()`s to identify the fragments you want to keep in
the orignally requested URL.  Use `$1`, `$2`, ... `$9` in the target URL
to include the captured fragments.

``` js
var config = {
  "forward": {
    "/remote-api/(.*)": "https://api.example.com/$1"
  }
};

```

This config would forward requests for `/remote-api/*` to
`https://api.example.biz:443/*`.


#### URL Rewriting to add a base path to the requested URL

The target server can prepend a base path for remote servers:

``` js
var config = {
  "forward": {
    "/junction/": "http://www.example.com/subapp"
  }
};

```

This config would forward requests for `/junction/*` to
`http://api.example.biz:80/subapp/junction/*`.


#### URL Rewriting to rearrange fragments

``` js
var config = {
  "forward": {
    "/user/(\\d+)/email/(\\S+)?(.*)": "http://api.example.com/account?id=$1&email=$2&$3"
  }
};

```

This config would forward requests for `/user/12345/email/987` to
`http://api.example.biz:443/account/12345/subscriptions/987`.


### Header Injection

The proxy can optionally inject headers into proxied requests.  This is useful
for remote endpoints that require headers values for authorization, like
API tokens.  The value of the injected header may be either a string value
or a function that accepts the request object and returns a string value.

Examples:
```
{
  proxy: {
    forward: {
      '/api/': 'http://api.example.com:8080'
    },
    headers: {
      'X-Forwarded-User': 'John Doe'
    }
  }
}
```

```
{
  proxy: {
    forward: {
      '/api/': 'http://api.example.com:8080'
    },
    headers: {
      'Authorization':  function(req) {
        accessToken = (req.user && req.user.access_token) ? req.user.access_token : 'api_token'
        return 'Bearer '.concat(accessToken)
      }
    }
  }
}
```

### WebSockets

WebSockets are not implemented yet.  WebSockets seem straightforward
to implement with the http-proxy module.
Please [create an issue](https://github.com/steve-jansen/json-proxy/issues/new) 
if you need WebSocket support, along some details on your desired use case to
help spec the tests.

### Developing

Unit tests are run with code coverage reporting via:

```
npm test
```

New features and fixes should include relevant test cases.

JSHint style checking should always pass:

```
npm run-script jshint
```

### Credits
This utility glues together the outstanding node packages 
[node-http-proxy by nodejitsu](https://github.com/nodejitsu/node-http-proxy) 
and [node-static by cloudhead](https://github.com/cloudhead/node-static) for 
proxying HTTP traffic and serving static files via HTTP.

Thanks to @ehtb for contributing new features.

### Issues
Please report bugs and features requets @ [https://github.com/steve-jansen/json-proxy/issues](https://github.com/steve-jansen/json-proxy/issues).


CHANGELOG.md
# Changelog
All notable changes to https://www.npmjs.org/package/json-proxy will be
documented in this file.

## 0.9.3

### Fixed
- Check for resource leak in proxy error callback

## 0.9.2

Dependency updates

## 0.9.1 - 2015-10-25

Nothing signficant.  Trigged by Travis w/ build updates.

## 0.9.0 - 2015-05-26

Release candidate.  The json-proxy has been stable for almost a year now, time
to consider rev'ing to 1.0.

### Added
- Nothing.

### Deprecated
- Nothing.

### Removed
- Git tagging with Travis. Will be done manually.

### Fixed
- Nothing.

## 0.4.0 - 2015-05-22

### Added
- Support for overrideing the `Host` and `Via` request headers.
- Continuous Delivery to npm via Travis CI

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- npm-shrinkwrap was out of date for v0.3.x
- Minor improvements to test suites structure


## 0.3.1 - 2015-03-18

### Added
- Nothing.

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Update dependencies
- Add Node 0.12 to Travis CI builds to verify compatability
- Minor documentation and style fixes


## 0.3.0 - 2014-08-01

### Added
- this CHANGELOG, thanks to http://keepachangelog.com/
- forwarding rules now support URL rewriting via nginx style regex captures
- injected headers can now use a function that accepts the req object and
  returns a simple string
- reuse the same unit test suites for proxying with and without a LAN HTTP proxy
- resolved codeclimate complexity warnings
- additional keywords in package.json for SEO with grunt plugins

### Deprecated
- Nothing.

### Removed
- Dropped support for parsing the v0.0.1 configuration format.  This means
  that top level configuration values must be inside a `server` or `proxy`
  configuration block.

### Fixed
- Restored express as a package dependency for `npm install -g json-proxy`.
- Fixed checks for `undefined` and `null` introduced by jshint fixes.


## 0.2.0 - 2014-07-15

### Added
- support for use as grunt middleware
- examples for running a proxy url inside `grunt serve`;
  see gruntjs/grunt-contrib-connect#85
- support for proxying to SSL/TLS endpoints
- support for basic authentication against HTTP proxy gateways;
  which is common on large LAN environments (e.g., corporations, universities);
  see v0.0.3
- internal refactoring to use the latest nodejitsu http-proxy bits
  (http-proxy@1.1.x); http-proxy@1.0 was a "from-scratch" implementation of
  the proxy core by nodejitsu, leaving http-proxy@0.x no longer supported
- general housekeeping: better unit test coverage, resolved jshint issues

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.


## 0.1.2  - 2014-01-25

### Added
- updated dependencies to use optimist@0.6.0

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- version of http-proxy dependency in package.json
- froze http-proxy dependency at last release prior to http-proxy@1.0,
  which introduced breaking API changes
- script path in the CLI example shell scripts

## 0.1.1  - 2013-10-21

### Added

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- typo in package.json


## 0.1.0  - 2013-08-01

### Added
- support for using json-proxy as express middleware
- CLI support for [AngularJS's html5mode](https://docs.angularjs.org/guide/$location)
- more robust configuration:
  - hierarchical fallback (from highest to lowest prioirty):
    - hard coded config values
    - command line options
    - config file options
    - environmental variables
    - json-proxy defaults
- proper support for `npm -g` installs
- mocha unit test and jshint linting
- examples for both CLI and middleware use cases
- an npm-shrinkwrap.json file

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

## 0.0.5  - 2013-04-02

### Added

### Deprecated
- CLI banner changed from deprecated product name `api-proxy` to `json-proxy`

### Removed
- Nothing.

### Fixed
- using the node-static middleware with NodeJS 0.10.*
- config file argument and file not found errors

## 0.0.3  - 2013-04-02

### Added
- support for storing config in json files to simplify CLI usage
- configuration files will replace the token `$config` with 
  the normalized path to the config file on disk
- support for custom headers, which is useful for endpoints that
  require reverse proxy authentication via headers
- support for connecting via HTTP proxy server gateways,
  which is common on large LAN environments (e.g., corporations, universities)

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.


## 0.0.1  - 2013-02-28

### Added
- alpha release as CLI utility
- a static file web server for UI developers that proxies JSON API calls to one
  or more remote web servers.

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.


lib/cli.js
#!/usr/bin/env node

'use strict';

var express = require('express'),
    morgan = require('morgan'),
    proxy = require('./proxy'),
    configParser = require('./config'),
    logger = require('./logger'),
    pkginfo = require('../package.json');


exports.run = function run() {
  // read the config file and/or commmand line parameters
  var argv = optimist();
  var app = express();
  var config = configParser({ file: argv.config, argv: argv });

  app.use(morgan('dev'));
  app.use(proxy.initialize({ file: argv.config, argv: argv }));
  app.use(express.static(config.server.webroot));
  if (config.server.html5mode) {
    // support AngularJS's html5 location mode,
    // which requires rewriting all requests for content
    // to the single page app entry point, which is usually index.html
    var target = (typeof(argv.html5mode) === "string") ? argv.html5mode : '/index.html';
    app.use(function(req, res){
      logger.info('html5', 'rewriting ', req.url , ' to ',  target);
      res.sendfile(config.server.webroot + target);
    });
  }

  welcome(config);

  app.listen(config.server.port, function() {
    logger.info('proxy', 'listening on port ' + config.server.port);
  });
};

function welcome(config) {
  logger.banner();

  logger.info('proxy', 'hosting local files from ' + config.server.webroot);

  if (undefined !== config.gateway && null !== config.gateway) {
    logger.info('proxy', 'routing forward rules via gateway ' + config.gateway.host + ':' + config.gateway.port);
  }

  config.proxy.forward.forEach(function(rule) {
    if (undefined !== rule && null !== rule) {
      logger.info('proxy', 'forwarding ' + rule.regexp.source + ' --> ' + rule.target.host + ':' + rule.target.port);
    }
 });

  config.proxy.headers.forEach(function(rule) {
    logger.info('proxy', 'forwarding injects header "' + rule.name+ ': ' + rule.value + '"');
  });

}

function optimist() {
  var usage = [
    pkginfo.name + ' v' + pkginfo.version,
    '',
    'A utility for UI developers to proxy JSON API calls to remote servers',
    'while serving all other requests locally without CORS or JSONP.',
    '',
    '',
    'Usage:',
    '  json-proxy [-c configFile] [-p port] [-f proxy forwarding rule]',
    '             [-h header rule] [-html5mode [defaultFile]] [directory]',
    '',
    'Examples:',
    '   json-proxy -p 8080 -f "/api=http://server" -f "/=http://localhost:9000" .',
    '   json-proxy -h "X-Forwarded-User=johndoe" /tmp/folder',
    '   json-proxy -c "/tmp/config.json"',
    '',
    'By default, looks for a config file at ./json-proxy.json',
    '',
    'Environmental variables:',
    '  JSON_PROXY_PORT         see --port',
    '  JSON_PROXY_WEBROOT      directory',
    '  JSON_PROXY_GATEWAY      --gateway',
    '  JSON_PROXY_GATEWAY_AUTH "username:password" credentials for --gateway)'
  ].join('\n');

  var argv = require('optimist')
             .usage(usage)
             .alias('p', 'port')
             .describe('p', 'The TCP port for the proxy server')
             .alias('f', 'forward')
             .describe('f', 'a forwarding rule (ex. /foo=server/foo)')
             .alias('h', 'header')
             .describe('h', 'a custom request header (ex. iv-user=johndoel)')
             .alias('c', 'config')
             .describe('c', 'a config file')
             .alias('g', 'gateway')
             .describe('g', 'URL for a LAN HTTP proxy to use for forwarding requests')
             .describe('html5mode', 'support AngularJS HTML5 mode by catching 404s')
             .alias('?', 'help')
             .describe('?', 'about this utility')
             .describe('version', 'version info')
             .argv;

  if (argv.help === true) {
    require('optimist').showHelp();
    process.exit(-1);
  }
  if (argv.version === true) {
    logger.info(pkginfo.name + ' v' + pkginfo.version);
    process.exit(0);
  }

  return argv;
}

